LoopKit/WeakObserverSet.swift

```diff
@@ -0,0 +1,147 @@
+//
+//  WeakObserverSet.swift
+//  LoopKit
+//
+//  Created by Michael Pangburn
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+
+/// A set-like collection of weak types, providing closure-baesd iteration on a client-specified queue
+/// Mutations and iterations are thread-safe
+public class WeakObserverSet<Observer> {
+    private typealias Identifier = ObjectIdentifier
+    private typealias Element = ObserverDispatchContainer<Observer>
+
+    private class ObserverDispatchContainer<Observer> {
+        private weak var _observer: AnyObject?
+        weak var queue: DispatchQueue?
+
+        var observer: Observer? {
+            return _observer as? Observer
+        }
+
+        init(observer: Observer, queue: DispatchQueue) {
+            // All Swift values are implicitly convertible to `AnyObject`,
+            // so this runtime check is the tradeoff for supporting class-constrained protocol types.
+            precondition(Mirror(reflecting: observer).displayStyle == .class, "Weak references can only be held of class types.")
+
+            self._observer = observer as AnyObject
+            self.queue = queue
+        }
+
+        func call(_ body: @escaping (_ observer: Observer) -> Void) {
+            guard let queue = self.queue, let observer = self.observer else {
+                return
+            }
+
+            queue.async {
+                body(observer)
+            }
+        }
+    }
+
+    private let observers: Locked<[Identifier: Element]>
+
+    public init() {
+        observers = Locked([:])
+    }
+
+    /// Adds an observer and its calling queue
+    ///
+    /// - Parameters:
+    ///   - observer: The observer
+    ///   - queue: The queue to use when calling the observer
+    public func addObserver(_ observer: Observer, queue: DispatchQueue) {
```

Comment:
Since we aren't observing the set itself, this function should be named `insert`.